### DISCUSSION TITLE
Inconsistent error throwing in Uri parser

### DISCUSSION THREAD

**Discussion by krypt-lynx** [resiprocate/resiprocate#223] [General]
Parsing error will be generated if malformed ipv6 address is encountered (example: '[1111:2222;garbage')
But, from look of things, there will be no error thrown if the string is not recognised as an ipv6 (the criteria is leading '[' symbol)

https://github.com/resiprocate/resiprocate/blob/7bc80ff508855a2304154d0ee00b6d1927574b08/resip/stack/Uri.cxx#L1109

In the wild, I encountered crash in third party library, caused by this behaviour (apparently they tested only ipv4 addresses)

**Comment by sgodin** [comment]
Hmmm - I figure the stack needs some mechanism to determine if some block of character is even an IPV6 address, and I suppose the leading '[' is our parsers key.  If the URI is still valid (according to the RFC's) without a leading '[', then I don't think there is a bug here.  Can you be more clear on what exact Uri is not throwing an error that you would expect to, and perhaps suggest how the stack might behave differently?

**Reply by krypt-lynx** [reply]
I have no ability to run tests of this sort, unfortunately. I got the app crash log from Firebase pointing on this method, and we found some malformed ipv6 in our logs.
From the looks of things there is no other exception throwing in whole Uri::parse and its used methods.

Looks like this condition expect address in format of `[FFFF:FFFF:FFFF:FFFF:FFFF]:9999` (ipv6 port declaration). 
This branch sends parts inside the brackets to `ParseBuffer::data` (or throws exception if data is garbage), and whole value otherwise (without any checks).

P.S.: Frankly, my C++ knowledge is on "read only" level.

**Reply by sgodin** [reply]
No worries on your C++ knowledge, I'm just asking general procedure questions.  Can you post the exact URI that you are expecting to fail parsing so I can see if it makes sense to make any changes?

**Reply by krypt-lynx** [reply]
From the looks of the code: `127.garbage` will fail without exception thrown.
I expect `[1:2garbage` and `1.2garbage` to fail in the same way. With exception thrown or not is different question
